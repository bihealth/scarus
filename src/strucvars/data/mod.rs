//! Representation and I/O of datasets.

pub mod clingen_dosage;
pub mod decipher_hi;
pub mod gnomad;
pub mod hgnc;
pub mod intervals;
