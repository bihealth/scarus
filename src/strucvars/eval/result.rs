//! Common code for evaluation results of CNVs.

use annonars::gnomad_sv::cli::query::Record as GnomadSvRecord;
use annonars::pbs::clinvar::sv::Record as ClinvarSvRecord;
use intervals_general::bound_pair::BoundPair;

use super::{del, dup};

/// Overall result.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
#[serde(rename_all = "snake_case")]
pub enum Evaluation {
    /// Result of assessing DUPs.
    Dup(dup::result::Evaluation),
    /// Result of assessing DELs.
    Del(del::result::Evaluation),
}

/// Overal assessment of a CNV.
#[derive(
    Debug,
    Clone,
    Copy,
    Default,
    PartialEq,
    strum_macros::EnumIter,
    serde::Deserialize,
    serde::Serialize,
)]
#[serde(rename_all = "snake_case")]
pub enum ClinicalSignificance {
    /// Benign
    Benign,
    /// Likely benign
    LikelyBenign,
    /// Uncertain_significance
    #[default]
    UncertainSignificance,
    /// Likely_pathogenic
    LikelyPathogenic,
    /// Pathogenic
    Pathogenic,
}

/// Score interval to use.
pub type ScoreRange = intervals_general::interval::Interval<f32>;

impl ClinicalSignificance {
    /// Return the scores,
    pub fn range(self) -> ScoreRange {
        match self {
            ClinicalSignificance::Benign => ScoreRange::UnboundedClosedRight { right: 0.99 },
            ClinicalSignificance::LikelyBenign => ScoreRange::LeftHalfOpen {
                bound_pair: BoundPair::new(-0.99, -0.90).expect("invalid interval?"),
            },
            ClinicalSignificance::UncertainSignificance => ScoreRange::Open {
                bound_pair: BoundPair::new(-0.90, 0.90).expect("invalid interval?"),
            },
            ClinicalSignificance::LikelyPathogenic => ScoreRange::LeftHalfOpen {
                bound_pair: BoundPair::new(0.90, 0.99).expect("invalid interval?"),
            },
            ClinicalSignificance::Pathogenic => ScoreRange::UnboundedClosedLeft { left: 0.99 },
        }
    }
}

/// Enumeration describing the PVS1 results.
#[derive(
    Debug,
    Clone,
    Copy,
    PartialEq,
    Eq,
    PartialOrd,
    Ord,
    Default,
    serde::Deserialize,
    serde::Serialize,
)]
#[serde(rename_all = "snake_case")]
pub enum Pvs1Result {
    /// PVS1
    #[default]
    Pvs1,
    /// PVS1_Strong
    Pvs1Strong,
    /// PVS1_Moderate
    Pvs1Moderate,
    /// PVS1_Supporting
    Pvs1Supporting,
}

/// Information about one overlapping ClinVar records.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct ClinvarSvOverlap {
    /// Reciprocal overlap.
    pub overlap: f32,
    /// Overlapping ClinVar record.
    pub record: ClinvarSvRecord,
}

/// Information about one overlapping gnomAD records.
#[derive(Debug, Clone, PartialEq, serde::Deserialize, serde::Serialize)]
pub struct GnomadSvOverlap {
    /// Reciprocal overlap.
    pub overlap: f32,
    /// Overlapping ClinVar record.
    pub record: GnomadSvRecord,
}
