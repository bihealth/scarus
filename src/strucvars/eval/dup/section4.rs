//! Implementation of evaluation of copy number loss section 4.
//!
//! Note that we can only determine 4O reliably in an automated fashion.  We support
//! reporting overlapping variants from ClinVar but leave them as "dangle" because a
//! human must evaluate the phenotype.

use super::result::{G4Patho, Section, G4, G4N};
use crate::strucvars::{
    ds::StructuralVariant,
    eval::{
        dup::result::G4O,
        result::{ClinvarSvOverlap, GnomadSvOverlap},
        CarrierFrequency, IntoInterval, Overlaps, SvType,
    },
};
use annonars::{
    clinvar_sv::cli::query::EXAC_CNV_CASES,
    gnomad_sv::cli::query::Record as GnomadSvRecord,
    pbs::clinvar::{
        minimal::{ClinicalSignificance, ReviewStatus, VariantType},
        sv::Record as ClinvarSvRecord,
    },
};
use bio::bio_types::genome::{AbstractInterval, Interval};

/// Evaluation of duplications, gain of copy number.
///
/// This is mainly used to encapsulate the functionality.  Creating new such
/// objects is very straightforward and cheap.
pub struct Evaluator<'a> {
    /// The parent evaluator.
    #[allow(dead_code)] // TODO
    parent: &'a super::super::Evaluator,
}

impl<'a> Evaluator<'a> {
    /// Create a new `Evaluator`.
    pub fn with_parent(parent: &'a super::super::Evaluator) -> Self {
        Self { parent }
    }

    /// Perform the evaluation of copy number loss Section 3 and all subsection.
    ///
    /// # Arguments
    ///
    /// * `strucvar` - Structural variant to be evaluated.
    ///
    /// # Returns
    ///
    /// The evaluation result.
    ///
    /// # Errors
    ///
    /// If anything goes wrong, it returns a generic `anyhow::Error`.
    pub fn evaluate(&self, strucvar: &StructuralVariant) -> Result<Vec<Section>, anyhow::Error> {
        tracing::debug!("evaluating section 4 for {:?}", strucvar);
        let sv_interval: Interval = strucvar.clone().into();
        let mut result = Vec::new();

        // Retrieve all overlapping ClinVar records.
        let clinvar_records = self.overlapping_clinvar_records(&sv_interval)?;

        if let Some(result_clinvar_patho) =
            self.handle_clinvar_pathogenic(&sv_interval, &clinvar_records)?
        {
            result.push(result_clinvar_patho)
        }
        if let Some(result_clinvar_benign) =
            self.handle_clinvar_benign(&sv_interval, &clinvar_records)?
        {
            result.push(result_clinvar_benign)
        }
        if let Some(result_gnomad_sv) = self.handle_gnomad_sv(&sv_interval)? {
            result.push(result_gnomad_sv)
        }

        tracing::warn!("Section 4 evaluation not implemented yet");
        Ok(result)
    }

    /// Obtain overlapping ClinVar records.
    fn overlapping_clinvar_records(
        &self,
        sv_interval: &Interval,
    ) -> Result<Vec<ClinvarSvRecord>, anyhow::Error> {
        tracing::debug!("querying ClinVar for {:?}", sv_interval);
        let res = self
            .parent
            .clinvar_sv_overlaps(
                sv_interval.contig(),
                sv_interval.range().start as u32,
                sv_interval.range().end as u32,
            )
            .map_err(|e| anyhow::anyhow!("ClinVar query failed: {}", e))?;
        tracing::debug!("found {} overlapping ClinVar records", res.len());
        tracing::trace!("records are {:#?}", &res);
        Ok(res)
    }

    /// Evaluate ClinVar pathogenic variants.
    ///
    /// This could be one of 4A, 4B, 4C, 4D, 4E, 4L, 4M.
    fn handle_clinvar_pathogenic(
        &self,
        sv_interval: &Interval,
        clinvar_records: &[ClinvarSvRecord],
    ) -> Result<Option<Section>, anyhow::Error> {
        let pathogenic = [
            ClinicalSignificance::Pathogenic as i32,
            ClinicalSignificance::LikelyPathogenic as i32,
        ];
        // Filter intervals to pathogenic ones without conflicts and sufficient overlap.
        let mut overlaps = clinvar_records
            .iter()
            .flat_map(|record| {
                let rc = record.reference_assertions.first().expect("no RCV");
                let record_interval = record.into_interval();
                let overlap = sv_interval.reciprocal_overlap(&record_interval) as f32;
                tracing::trace!("considering records {:?} with overlap {}", record, overlap);
                tracing::trace!("rc.clinical_significance = {} / {:?}", rc.clinical_significance, &pathogenic);
                let is_pathogenic = pathogenic.contains(&rc.clinical_significance);
                let is_review_status = rc.review_status <= ReviewStatus::CriteriaProvidedSingleSubmitter as i32;
                let is_variant_type = record.variant_type == VariantType::Duplication as i32;
                let is_overlap = overlap >= self.parent.config.clinvar_sv_min_overlap_pathogenic;
                let is_interesting = is_pathogenic && is_variant_type && is_overlap;
                tracing::trace!("is_pathogenic: {}, is_review_status: {}, is_variant_type: {}, is_overlap: {}, is_interesting: {}", is_pathogenic, is_review_status, is_variant_type, is_overlap, is_interesting);
                if is_interesting {
                    Some(ClinvarSvOverlap {
                        overlap,
                        record: record.clone(),
                    })
                } else {
                    None
                }
            })
            .collect::<Vec<_>>();

        overlaps.sort_by(|a, b| {
            b.overlap
                .partial_cmp(&a.overlap)
                .unwrap_or(std::cmp::Ordering::Equal)
        });

        if overlaps.is_empty() {
            Ok(None)
        } else {
            Ok(Some(Section::G4(G4::G4Patho(G4Patho { overlaps }))))
        }
    }

    /// Evaluate ClinVar benign variants.
    ///
    /// This is roughly 4N.
    fn handle_clinvar_benign(
        &self,
        sv_interval: &Interval,
        clinvar_records: &[ClinvarSvRecord],
    ) -> Result<Option<Section>, anyhow::Error> {
        let benign = [
            ClinicalSignificance::Benign as i32,
            ClinicalSignificance::LikelyBenign as i32,
        ];
        // Filter intervals to pathogenic ones without conflicts and sufficient overlap.
        let mut overlaps = clinvar_records
            .iter()
            .flat_map(|record| {
                let rc = record.reference_assertions.first().expect("no RCV");
                let record_interval = record.into_interval();
                let overlap = sv_interval.reciprocal_overlap(&record_interval) as f32;
                tracing::trace!("considering records {:?} with overlap {}", record, overlap);
                let interesting = benign.contains(&rc.clinical_significance)
                    && rc.review_status <= ReviewStatus::CriteriaProvidedSingleSubmitter as i32
                    && record.variant_type == VariantType::Duplication as i32
                    && overlap >= self.parent.config.clinvar_sv_min_overlap_pathogenic;
                if interesting {
                    Some(ClinvarSvOverlap {
                        overlap,
                        record: record.clone(),
                    })
                } else {
                    None
                }
            })
            .collect::<Vec<_>>();

        overlaps.sort_by(|a, b| {
            b.overlap
                .partial_cmp(&a.overlap)
                .unwrap_or(std::cmp::Ordering::Equal)
        });

        if overlaps.is_empty() {
            Ok(None)
        } else {
            Ok(Some(Section::G4(G4::G4N(G4N { overlaps }))))
        }
    }

    /// Handle overlap in gnomAD-SV.
    ///
    /// This corresponds to case 4O.
    fn handle_gnomad_sv(&self, sv_interval: &Interval) -> Result<Option<Section>, anyhow::Error> {
        // Query gnomAD-SV for all records.
        tracing::debug!("querying gnomAD-SV for {:?}", sv_interval);
        let raw_records = self
            .parent
            .gnomad_sv_overlaps(
                sv_interval.contig(),
                sv_interval.range().start as u32,
                sv_interval.range().end as u32,
            )
            .map_err(|e| anyhow::anyhow!("gnomAD-SV query failed: {}", e))?;
        tracing::debug!("found {} overlapping gnomAD-SV records", raw_records.len());
        tracing::trace!("records are {:#?}", &raw_records);

        // Filter to records with the correct type and appropriate overlap.
        let mut overlaps: Vec<GnomadSvOverlap> = raw_records
            .into_iter()
            .flat_map(|record| {
                let record_interval = record.into_interval();
                let overlap = sv_interval.reciprocal_overlap(&record_interval) as f32;
                let sv_type = SvType::try_from(&record);
                let interesting = matches!(sv_type, Ok(SvType::Dup | SvType::Cnv))
                    && overlap >= self.parent.config.gnomad_sv_min_overlap;
                if interesting {
                    Some(GnomadSvOverlap { overlap, record })
                } else {
                    None
                }
            })
            .collect::<Vec<_>>();
        overlaps.sort_by(|a, b| {
            b.overlap
                .partial_cmp(&a.overlap)
                .unwrap_or(std::cmp::Ordering::Equal)
        });
        tracing::trace!("filtered overlaps are {:#?}", &overlaps);

        // We have to manually count ExAC CNV records as they are per-sample and not aggregated.
        let exac_cnv_count = overlaps
            .iter()
            .filter(|GnomadSvOverlap { record, .. }| matches!(record, &GnomadSvRecord::ExacCnv(_)))
            .count();
        let exac_freq = exac_cnv_count as f32 / EXAC_CNV_CASES as f32;
        // For the rest, we can obtain the frequency from the records.
        let other_freqs = overlaps
            .iter()
            .flat_map(|GnomadSvOverlap { record, .. }| {
                if matches!(record, &GnomadSvRecord::ExacCnv(_)) {
                    None
                } else {
                    Some(record.carrier_freq().expect("no carrier freq?"))
                }
            })
            .collect::<Vec<_>>();
        let max_freq = other_freqs
            .iter()
            .max_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal))
            .copied()
            .unwrap_or(0.0);

        let suggested_score = if exac_freq >= self.parent.config.gnomad_sv_min_freq_common
            || max_freq >= self.parent.config.gnomad_sv_min_freq_common
        {
            -1.0
        } else {
            0.0
        };

        Ok(if overlaps.is_empty() {
            None
        } else {
            Some(Section::G4(G4::G4O(G4O {
                suggested_score,
                overlaps,
            })))
        })
    }
}

#[cfg(test)]
mod test {
    use crate::strucvars::ds::{StructuralVariant, SvType};

    use super::super::super::test::global_evaluator_37;

    #[tracing_test::traced_test]
    #[rstest::rstest]
    // ClinVar pathogenic variant
    #[case("10", 55_581_618, 56_288_162, "VCV000806586")]
    // ClinVar benign variant
    #[case("10", 135_252_347, 135_378_802, "VCV000221841")]
    // gnomAD SV variants
    // ExAC CNV
    #[case("10", 92_994, 95_183, "chr10-92994-95183-NFE")]
    // gnomAD-SV v2
    #[case("1", 157_000, 166_000, "gnomAD-SV_v2.1_DUP_1_6")]
    fn test_evaluate(
        #[case] chrom: &str,
        #[case] start: u32,
        #[case] stop: u32,
        #[case] label: &str,
        global_evaluator_37: &super::super::super::Evaluator,
    ) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{}", label);

        let evaluator = super::Evaluator::with_parent(global_evaluator_37);

        let strucvar = StructuralVariant {
            chrom: chrom.into(),
            start,
            stop,
            svtype: SvType::Dup,
            ambiguous_range: None,
        };

        insta::assert_yaml_snapshot!(evaluator.evaluate(&strucvar)?);

        Ok(())
    }

    #[tracing_test::traced_test]
    #[rstest::rstest]
    // same as in `evaluate` test above, keep in sync
    #[case("10", 55_581_618, 56_288_162, "VCV000806586")]
    fn handle_clinvar_pathogenic(
        #[case] chrom: &str,
        #[case] start: u32,
        #[case] stop: u32,
        #[case] label: &str,
        global_evaluator_37: &super::super::super::Evaluator,
    ) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{}", label);

        let evaluator = super::Evaluator::with_parent(global_evaluator_37);
        let strucvar = StructuralVariant {
            chrom: chrom.into(),
            start,
            stop,
            svtype: SvType::Dup,
            ambiguous_range: None,
        };
        let sv_interval = strucvar.into();

        let clinvar_records = evaluator.overlapping_clinvar_records(&sv_interval)?;

        insta::assert_yaml_snapshot!(
            evaluator.handle_clinvar_pathogenic(&sv_interval, &clinvar_records)?
        );

        Ok(())
    }

    #[tracing_test::traced_test]
    #[rstest::rstest]
    // same as in `evaluate` test above, keep in sync
    #[case("10", 135_252_347, 135_378_802, "VCV000221841")]
    fn handle_clinvar_benign(
        #[case] chrom: &str,
        #[case] start: u32,
        #[case] stop: u32,
        #[case] label: &str,
        global_evaluator_37: &super::super::super::Evaluator,
    ) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{}", label);

        let evaluator = super::Evaluator::with_parent(global_evaluator_37);
        let strucvar = StructuralVariant {
            chrom: chrom.into(),
            start,
            stop,
            svtype: SvType::Dup,
            ambiguous_range: None,
        };
        let sv_interval = strucvar.into();

        let clinvar_records = evaluator.overlapping_clinvar_records(&sv_interval)?;

        insta::assert_yaml_snapshot!(
            evaluator.handle_clinvar_benign(&sv_interval, &clinvar_records)?
        );

        Ok(())
    }

    #[tracing_test::traced_test]
    #[rstest::rstest]
    // same as in `evaluate` test above, keep in sync
    // ExAC CNV
    #[case("10", 92_994, 95_183, "chr10-92994-95183-NFE")]
    // gnomAD-SV v2
    #[case("1", 157_000, 166_000, "gnomAD-SV_v2.1_DUP_1_6")]
    fn handle_gnomad_sv(
        #[case] chrom: &str,
        #[case] start: u32,
        #[case] stop: u32,
        #[case] label: &str,
        global_evaluator_37: &super::super::super::Evaluator,
    ) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!("{}", label);

        let evaluator = super::Evaluator::with_parent(global_evaluator_37);
        let strucvar = StructuralVariant {
            chrom: chrom.into(),
            start,
            stop,
            svtype: SvType::Dup,
            ambiguous_range: None,
        };
        let sv_interval = strucvar.into();

        insta::assert_yaml_snapshot!(evaluator.handle_gnomad_sv(&sv_interval)?);

        Ok(())
    }
}
