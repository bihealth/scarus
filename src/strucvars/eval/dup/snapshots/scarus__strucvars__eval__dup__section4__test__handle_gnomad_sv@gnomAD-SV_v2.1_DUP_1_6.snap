---
source: src/strucvars/eval/dup/section4.rs
expression: evaluator.handle_gnomad_sv(&sv_interval)?
---
g4:
  g4o:
    suggested_score: -1
    overlaps:
      - overlap: 1
        record:
          gnomad_sv2:
            chrom: "1"
            pos: 157000
            end: 166000
            id: gnomAD-SV_v2.1_DUP_1_6
            filters:
              - FILTER_LOW_CALL_RATE
              - FILTER_UNSTABLE_AF_PCRMINUS
            svType: SV_TYPE_DUP
            alleleCounts:
              - bySex:
                  overall:
                    ac: 384
                    an: 13770
                    af: 0.027887
                    nBiGenos: 6885
                    nHomref: 6505
                    nHet: 376
                    nHomalt: 4
                    freqHomref: 0.944808
                    freqHet: 0.0546115
                    freqHomalt: 0.000580973
                  xx:
                    ac: 167
                    an: 6902
                    af: 0.024196
                    nBiGenos: 3451
                    nHomref: 3285
                    nHet: 165
                    nHomalt: 1
                    freqHomref: 0.951898
                    freqHet: 0.0478122
                    freqHomalt: 0.000289771
                  xy:
                    ac: 214
                    an: 6834
                    af: 0.031314
                    nBiGenos: 3417
                    nHomref: 3206
                    nHet: 208
                    nHomalt: 3
                    freqHomref: 0.93825
                    freqHet: 0.0608721
                    freqHomalt: 0.000877963
                byPopulation:
                  - population: POPULATION_AFR
                    counts:
                      overall:
                        ac: 256
                        an: 5434
                        af: 0.047111
                        nBiGenos: 2717
                        nHomref: 2464
                        nHet: 250
                        nHomalt: 3
                        freqHomref: 0.906883
                        freqHet: 0.0920132
                        freqHomalt: 0.00110416
                      xx:
                        ac: 113
                        an: 2480
                        af: 0.045565
                        nBiGenos: 1240
                        nHomref: 1128
                        nHet: 111
                        nHomalt: 1
                        freqHomref: 0.909677
                        freqHet: 0.0895161
                        freqHomalt: 0.000806452
                      xy:
                        ac: 141
                        an: 2946
                        af: 0.047862
                        nBiGenos: 1473
                        nHomref: 1334
                        nHet: 137
                        nHomalt: 2
                        freqHomref: 0.905635
                        freqHet: 0.0930075
                        freqHomalt: 0.00135777
                  - population: POPULATION_AMR
                    counts:
                      overall:
                        ac: 23
                        an: 970
                        af: 0.023711
                        nBiGenos: 485
                        nHomref: 462
                        nHet: 23
                        freqHomref: 0.952577
                        freqHet: 0.0474227
                      xx:
                        ac: 9
                        an: 522
                        af: 0.017241
                        nBiGenos: 261
                        nHomref: 252
                        nHet: 9
                        freqHomref: 0.965517
                        freqHet: 0.0344828
                      xy:
                        ac: 14
                        an: 446
                        af: 0.03139
                        nBiGenos: 223
                        nHomref: 209
                        nHet: 14
                        freqHomref: 0.93722
                        freqHet: 0.0627803
                  - population: POPULATION_EAS
                    counts:
                      overall:
                        ac: 71
                        an: 1366
                        af: 0.051977
                        nBiGenos: 683
                        nHomref: 613
                        nHet: 69
                        nHomalt: 1
                        freqHomref: 0.897511
                        freqHet: 0.101025
                        freqHomalt: 0.00146413
                      xx:
                        ac: 24
                        an: 576
                        af: 0.041667
                        nBiGenos: 288
                        nHomref: 264
                        nHet: 24
                        freqHomref: 0.916667
                        freqHet: 0.0833333
                      xy:
                        ac: 46
                        an: 784
                        af: 0.058673
                        nBiGenos: 392
                        nHomref: 347
                        nHet: 44
                        nHomalt: 1
                        freqHomref: 0.885204
                        freqHet: 0.112245
                        freqHomalt: 0.00255102
                  - population: POPULATION_EUR
                    counts:
                      overall:
                        ac: 31
                        an: 5874
                        af: 0.005277
                        nBiGenos: 2937
                        nHomref: 2906
                        nHet: 31
                        freqHomref: 0.989445
                        freqHet: 0.010555
                      xx:
                        ac: 18
                        an: 3244
                        af: 0.005549
                        nBiGenos: 1622
                        nHomref: 1604
                        nHet: 18
                        freqHomref: 0.988903
                        freqHet: 0.0110974
                      xy:
                        ac: 13
                        an: 2612
                        af: 0.004977
                        nBiGenos: 1306
                        nHomref: 1293
                        nHet: 13
                        freqHomref: 0.990046
                        freqHet: 0.00995406
                  - population: POPULATION_OTHER
                    counts:
                      overall:
                        ac: 3
                        an: 126
                        af: 0.02381
                        nBiGenos: 63
                        nHomref: 60
                        nHet: 3
                        freqHomref: 0.952381
                        freqHet: 0.047619
                      xx:
                        ac: 3
                        an: 80
                        af: 0.0375
                        nBiGenos: 40
                        nHomref: 37
                        nHet: 3
                        freqHomref: 0.925
                        freqHet: 0.075
                      xy:
                        an: 46
                        nBiGenos: 23
                        nHomref: 23
                        freqHomref: 1
              - cohort: controls
                bySex:
                  overall:
                    ac: 184
                    an: 6596
                    af: 0.027896
                    nBiGenos: 3298
                    nHomref: 3116
                    nHet: 180
                    nHomalt: 2
                    freqHomref: 0.944815
                    freqHet: 0.0545785
                    freqHomalt: 0.000606428
                  xx:
                    ac: 86
                    an: 3356
                    af: 0.025626
                    nBiGenos: 1678
                    nHomref: 1592
                    nHet: 86
                    freqHomref: 0.948749
                    freqHet: 0.0512515
                  xy:
                    ac: 97
                    an: 3228
                    af: 0.03005
                    nBiGenos: 1614
                    nHomref: 1519
                    nHet: 93
                    nHomalt: 2
                    freqHomref: 0.94114
                    freqHet: 0.0576208
                    freqHomalt: 0.00123916
                byPopulation:
                  - population: POPULATION_AFR
                    counts:
                      overall:
                        ac: 98
                        an: 2326
                        af: 0.042132
                        nBiGenos: 1163
                        nHomref: 1066
                        nHet: 96
                        nHomalt: 1
                        freqHomref: 0.916595
                        freqHet: 0.0825451
                        freqHomalt: 0.000859845
                      xx:
                        ac: 51
                        an: 1178
                        af: 0.043294
                        nBiGenos: 589
                        nHomref: 538
                        nHet: 51
                        freqHomref: 0.913413
                        freqHet: 0.0865874
                      xy:
                        ac: 47
                        an: 1148
                        af: 0.040941
                        nBiGenos: 574
                        nHomref: 528
                        nHet: 45
                        nHomalt: 1
                        freqHomref: 0.919861
                        freqHet: 0.0783972
                        freqHomalt: 0.00174216
                  - population: POPULATION_AMR
                    counts:
                      overall:
                        ac: 19
                        an: 660
                        af: 0.028788
                        nBiGenos: 330
                        nHomref: 311
                        nHet: 19
                        freqHomref: 0.942424
                        freqHet: 0.0575758
                      xx:
                        ac: 8
                        an: 362
                        af: 0.022099
                        nBiGenos: 181
                        nHomref: 173
                        nHet: 8
                        freqHomref: 0.955801
                        freqHet: 0.0441989
                      xy:
                        ac: 11
                        an: 296
                        af: 0.037162
                        nBiGenos: 148
                        nHomref: 137
                        nHet: 11
                        freqHomref: 0.925676
                        freqHet: 0.0743243
                  - population: POPULATION_EAS
                    counts:
                      overall:
                        ac: 53
                        an: 964
                        af: 0.054979
                        nBiGenos: 482
                        nHomref: 430
                        nHet: 51
                        nHomalt: 1
                        freqHomref: 0.892116
                        freqHet: 0.105809
                        freqHomalt: 0.00207469
                      xx:
                        ac: 19
                        an: 430
                        af: 0.044186
                        nBiGenos: 215
                        nHomref: 196
                        nHet: 19
                        freqHomref: 0.911628
                        freqHet: 0.0883721
                      xy:
                        ac: 33
                        an: 528
                        af: 0.0625
                        nBiGenos: 264
                        nHomref: 232
                        nHet: 31
                        nHomalt: 1
                        freqHomref: 0.878788
                        freqHet: 0.117424
                        freqHomalt: 0.00378788
                  - population: POPULATION_EUR
                    counts:
                      overall:
                        ac: 13
                        an: 2578
                        af: 0.005043
                        nBiGenos: 1289
                        nHomref: 1276
                        nHet: 13
                        freqHomref: 0.989915
                        freqHet: 0.0100853
                      xx:
                        ac: 7
                        an: 1344
                        af: 0.005208
                        nBiGenos: 672
                        nHomref: 665
                        nHet: 7
                        freqHomref: 0.989583
                        freqHet: 0.0104167
                      xy:
                        ac: 6
                        an: 1230
                        af: 0.004878
                        nBiGenos: 615
                        nHomref: 609
                        nHet: 6
                        freqHomref: 0.990244
                        freqHet: 0.0097561
                  - population: POPULATION_OTHER
                    counts:
                      overall:
                        ac: 1
                        an: 68
                        af: 0.014706
                        nBiGenos: 34
                        nHomref: 33
                        nHet: 1
                        freqHomref: 0.970588
                        freqHet: 0.0294118
                      xx:
                        ac: 1
                        an: 42
                        af: 0.02381
                        nBiGenos: 21
                        nHomref: 20
                        nHet: 1
                        freqHomref: 0.952381
                        freqHet: 0.047619
                      xy:
                        an: 26
                        nBiGenos: 13
                        nHomref: 13
                        freqHomref: 1
              - cohort: non_neuro
                bySex:
                  overall:
                    ac: 258
                    an: 11128
                    af: 0.023185
                    nBiGenos: 5564
                    nHomref: 5308
                    nHet: 254
                    nHomalt: 2
                    freqHomref: 0.95399
                    freqHet: 0.0456506
                    freqHomalt: 0.000359454
                  xx:
                    ac: 125
                    an: 5936
                    af: 0.021058
                    nBiGenos: 2968
                    nHomref: 2843
                    nHet: 125
                    freqHomref: 0.957884
                    freqHet: 0.0421159
                  xy:
                    ac: 132
                    an: 5166
                    af: 0.025552
                    nBiGenos: 2583
                    nHomref: 2453
                    nHet: 128
                    nHomalt: 2
                    freqHomref: 0.949671
                    freqHet: 0.0495548
                    freqHomalt: 0.000774293
                byPopulation:
                  - population: POPULATION_AFR
                    counts:
                      overall:
                        ac: 130
                        an: 2910
                        af: 0.044674
                        nBiGenos: 1455
                        nHomref: 1326
                        nHet: 128
                        nHomalt: 1
                        freqHomref: 0.91134
                        freqHet: 0.0879725
                        freqHomalt: 0.000687285
                      xx:
                        ac: 71
                        an: 1584
                        af: 0.044823
                        nBiGenos: 792
                        nHomref: 721
                        nHet: 71
                        freqHomref: 0.910354
                        freqHet: 0.0896465
                      xy:
                        ac: 59
                        an: 1326
                        af: 0.044495
                        nBiGenos: 663
                        nHomref: 605
                        nHet: 57
                        nHomalt: 1
                        freqHomref: 0.912519
                        freqHet: 0.0859729
                        freqHomalt: 0.0015083
                  - population: POPULATION_AMR
                    counts:
                      overall:
                        ac: 23
                        an: 970
                        af: 0.023711
                        nBiGenos: 485
                        nHomref: 462
                        nHet: 23
                        freqHomref: 0.952577
                        freqHet: 0.0474227
                      xx:
                        ac: 9
                        an: 522
                        af: 0.017241
                        nBiGenos: 261
                        nHomref: 252
                        nHet: 9
                        freqHomref: 0.965517
                        freqHet: 0.0344828
                      xy:
                        ac: 14
                        an: 446
                        af: 0.03139
                        nBiGenos: 223
                        nHomref: 209
                        nHet: 14
                        freqHomref: 0.93722
                        freqHet: 0.0627803
                  - population: POPULATION_EAS
                    counts:
                      overall:
                        ac: 71
                        an: 1366
                        af: 0.051977
                        nBiGenos: 683
                        nHomref: 613
                        nHet: 69
                        nHomalt: 1
                        freqHomref: 0.897511
                        freqHet: 0.101025
                        freqHomalt: 0.00146413
                      xx:
                        ac: 24
                        an: 576
                        af: 0.041667
                        nBiGenos: 288
                        nHomref: 264
                        nHet: 24
                        freqHomref: 0.916667
                        freqHet: 0.0833333
                      xy:
                        ac: 46
                        an: 784
                        af: 0.058673
                        nBiGenos: 392
                        nHomref: 347
                        nHet: 44
                        nHomalt: 1
                        freqHomref: 0.885204
                        freqHet: 0.112245
                        freqHomalt: 0.00255102
                  - population: POPULATION_EUR
                    counts:
                      overall:
                        ac: 31
                        an: 5756
                        af: 0.005386
                        nBiGenos: 2878
                        nHomref: 2847
                        nHet: 31
                        freqHomref: 0.989229
                        freqHet: 0.0107714
                      xx:
                        ac: 18
                        an: 3174
                        af: 0.005671
                        nBiGenos: 1587
                        nHomref: 1569
                        nHet: 18
                        freqHomref: 0.988658
                        freqHet: 0.0113422
                      xy:
                        ac: 13
                        an: 2564
                        af: 0.00507
                        nBiGenos: 1282
                        nHomref: 1269
                        nHet: 13
                        freqHomref: 0.98986
                        freqHet: 0.0101404
                  - population: POPULATION_OTHER
                    counts:
                      overall:
                        ac: 3
                        an: 126
                        af: 0.02381
                        nBiGenos: 63
                        nHomref: 60
                        nHet: 3
                        freqHomref: 0.952381
                        freqHet: 0.047619
                      xx:
                        ac: 3
                        an: 80
                        af: 0.0375
                        nBiGenos: 40
                        nHomref: 37
                        nHet: 3
                        freqHomref: 0.925
                        freqHet: 0.075
                      xy:
                        an: 46
                        nBiGenos: 23
                        nHomref: 23
                        freqHomref: 1

